### PR TITLE
AO3-5153: New approach to fixing auth tokens dynamically

### PR DESCRIPTION
**Pull request by elzj**
## Issue

https://otwarchive.atlassian.net/browse/AO3-5153

## Purpose

Dynamically updates form authenticity tokens for logged out users, to get around page caching bugs.

## Testing

You should be able to log in on staging.

### CODE REVIEW

**Comment by zz9pzza** [comment]
Don't we still need the cross site forgery protection even for guests ?

**Comment by elzj** [comment]
@zz9pzza This still uses the user's session to get their own form tokens, it doesn't actually bypass security.

**Comment by zz9pzza** [comment]
Can we do the bit where the javascript is only used for guests ?

**Comment by elzj** [comment]
It's only invoked if it sees the login form, which means you're logged out.

File: public/javascripts/application.js [hunk 1]
```diff
@@ -643,3 +644,18 @@ function thermometer() {
...
+  // we only do full page caching when users are logged out
+  if ($j('#small_login').length > 0) {
+    $j.getJSON("/token_dispenser.json", function( data ) {
+      var token = data.token;
+      //set token on fields
+      $j('input[name=authenticity_token]').each(function(){
+        $j(this).attr('value', token);
+      });
+      $j('meta[name=csrf-token]').attr('value', token);
+      console.log("I got a token: " + token);
```
Comment (zz9pzza) [inline public/javascripts/application.js]: Is this a good idea to deploy ?

**Pull request merged**